src/methods/join.js

```diff
@@ -589,3 +589,44 @@ export function joinBooleans(v1: void | boolean, v2: void | boolean): void | boo
     return v1 || v2;
   }
 }
+
+// Creates a single map that intersects maps m1 and m2 using the given intersect operator for values.
+export function intersectMaps<K, V>(
+  m1: Map<K, void | V>,
+  m2: Map<K, void | V>,
+  intersect: (K, void | V, void | V) => V
+): Map<K, void | V> {
+  let m3: Map<K, void | V> = new Map();
+  m1.forEach((val1, key, map1) => {
+    if (!m2.has(key)) return;
+    let val2 = m2.get(key);
+    let val3 = intersect(key, val1, val2);
+    m3.set(key, val3);
+  });
+  return m3;
+}
+
+export function intersectEffects(realm: Realm, e1: Effects, e2: Effects): Effects {
+  let [c1, g1, b1, p1, o1] = e1;
+  let [c2, g2, b2, p2, o2] = e2;
+
+  // Neither set of effects is allowed to be forked or abrupt
+  invariant(c1 instanceof Value && c2 instanceof Value);
+
+  // ignore generators
+  g1;
+  g2;
+  // intersect bindings
+  let conflict = realm.intrinsics.empty;
+  let b3 = intersectMaps(b1, b2, (b, v1, v2) => (v1 === v2 ? v1 : conflict));
+  let eqVal = (v1, v2) =>
+    v1 === v2 ||
+    (v1 instanceof ConcreteValue && v2 instanceof ConcreteValue && StrictEqualityComparison(realm, v1, v2));
+  let eqDesc = (d1, d2) => d1 && d2 && equalDescriptors(d1, d2) && eqVal(d1.value, d2.value);
+  let p3 = intersectMaps(p1, p2, (b, d1, d2) => (eqDesc(d1, d2) ? d1 : { value: conflict }));
+  // ignore created objects
+  o1;
+  o2;
+
+  return [realm.intrinsics.empty, new Generator(realm), b3, p3, new Set()];
```

Comment:
Judging by the function name `intersectEffects`, I find `realm.intrinsics.empty` and perhaps also `new Generator(realm)` surprising. Expected would be to at least check if c1 === c2, and then return c1 else empty, or probably better, generating code that does it dynamically. Similarly, deep equality check for generator.

I understand that this isn't needed for the concrete scenario at hand, but at least one prominent comment would be good.